### ISSUE TITLE
Audit module

### ISSUE THREAD

**Issue by ddeboer** [zotonic/zotonic#1529]
Continued from https://github.com/zotonic/zotonic/issues/1182#issuecomment-261201253.

> What about the amount of time the actions need to be stored? Two months seems reasonable.

Why not forever? In the [event sourcing](http://martinfowler.com/eaaDev/EventSourcing.html) way of building applications, every change is captured and persisted. The idea is to store the events forever to keep a full history of all changes. Why not do the same?

**Comment by mmzeeman** [comment]
Because an audit log can contains privacy sensitive information which I don't want to keep on a server forever.

**Comment by ddeboer** [comment]
You can always anonymise event details.

**Comment by mmzeeman** [comment]
That is why I want to be able to export the data. In that way the owners the audit data can take care of the privacy risk themselves. That is not my problem.

**Comment by mworrell** [comment]
I agree - store on the server the real non-anonymized data for max 2 months.

We can have export of events to an 3rd system for aggregation and statistical analysis. As the events are probably notifications, an extra module could just copy them over at the moment it happens.

**Comment by ddeboer** [comment]
I started work on a simple `mod_audit` that observes data changes and then persists them in the log table.

**Comment by mworrell** [comment]
I was thinking that we might use lager metadata to direct some log entries to the admin-accessible log.

Actually, that log might also be on disk and rotated?

**Comment by mmzeeman** [comment]
We have an audit module in channel. Can we join forces somehow?

**Comment by mmzeeman** [comment]
It logs audit events in a separate audit table. It logs an event category, the user-id which initiated it, the content group, and an associated timestamp.

**Comment by ddeboer** [comment]
Let’s work together on #2042! Comments are very welcome there.

**Comment by mmzeeman** [comment]
Ours is here: https://github.com/channelme/mod_admin_audit

**Comment by mmzeeman** [comment]
We build this module for one of our clients which also requested the possibility to query the data.
